### ISSUE TITLE
(3) CLI does not exit with non zero when playbook execution fails

### ISSUE THREAD

**Issue by rywillia** [CentOS-PaaS-SIG/linchpin#306]
I am running Linchpin version 1.0.1 and the exit code when my Ansible playbook fails is 0. I would expect this code to be 1 or another integer value. Having this would be great for automation being able to determine if my command executed successfully or not.

```
--snippet--
TASK [openstack : Initiating  Provisioning/Deprovioning of resources Openstack resource group] ***
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_resource_group.yml for localhost

TASK [openstack : Unset the authvar from previous run] *************************
ok: [localhost]

TASK [openstack : set default_cred_profile when res_grp[credentials] is undefined] ***
skipping: [localhost]

TASK [openstack : set default_cred_profile when res_grp[credentials] is undefined] ***
skipping: [localhost]

TASK [openstack : set cred_profile when res_grp[credentials] is defined] *******
ok: [localhost]

TASK [openstack : set default_cred_filename when res_grp[credentials] is defined] ***
ok: [localhost]

TASK [openstack : Get creds from auth driver] **********************************
changed: [localhost]

TASK [openstack : set auth_var] ************************************************
ok: [localhost]

TASK [openstack : provisioning resource definitions of current group] **********
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_res_defs.yml for localhost

TASK [openstack : provision resources of type os_server] ***********************
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_os_server.yml for localhost

TASK [openstack : provisioning os_server resources with provided auth] *********
fatal: [localhost]: FAILED! => {"changed": false, "extra_data": null, "failed": true, "msg": "Timeout waiting for the floating IP to be attached."}
	to retry, use: --limit @/usr/lib/python2.7/site-packages/linchpin/provision/site.retry

PLAY RECAP *********************************************************************
localhost                  : ok=18   changed=1    unreachable=0    failed=1   

hook postup initiated
[root@91b51ebff31f linchpin]# echo $?
0
```
I noticed there was a similar issue to this #159. This bug may still exist for certain use cases in the latest release?

**Comment by herlo** [comment]
It indeed looks like this bug still exists. I do have exit codes in place for most situations. Can you provide me with your workspace data (don't include credentials if they are in the workspace) along with the command and full output of what has happened? I'd like to replicate this problem.

**Comment by rywillia** [comment]
I am able to reproduce this problem in various situations.

1. Perform a linchpin init and immediately run linchpin -v up (using the example topology, layout, PinFile).
```
[root@1f1ad66a679f lp]# linchpin -v up
ctx.workspace: /home/lp
from_api: True
calling: preup
hook preup initiated

PLAY [schema check and Pre Provisioning Activities on topology_file] ***********

TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]

TASK [common : set lp_path] ****************************************************
skipping: [localhost]

TASK [common : set lpconfig (linchpin.conf) full path] *************************
skipping: [localhost]

TASK [common : set boolean evars from linchpin.conf] ***************************
skipping: [localhost]

TASK [common : set schema evars from linchpin.conf] ****************************
skipping: [localhost]

TASK [common : set default path evars from linchpin.conf] **********************
skipping: [localhost]

TASK [common : set remaining evars from linchpin.conf] *************************
skipping: [localhost]

TASK [common : set default path evars using workspace var] *********************
skipping: [localhost]

TASK [common : if topology_file is passed, set topology to its value] **********
skipping: [localhost]

TASK [common : if schema_file is passed, set schema to its value] **************
skipping: [localhost]

TASK [common : check topology file against the provided schema] ****************
skipping: [localhost]

TASK [common : assign output from provided schema] *****************************
skipping: [localhost]

TASK [common : check topology file against schema_v3] **************************
ok: [localhost]

TASK [common : Set state for schema_v3 validation] *****************************
ok: [localhost]

TASK [common : assign schema_v3 output] ****************************************
ok: [localhost]

TASK [common : check topology file against schema_v4] **************************
skipping: [localhost]

TASK [common : update output when schema_v4 is used] ***************************
skipping: [localhost]

TASK [common : assign output from schema_v4] ***********************************
skipping: [localhost]

TASK [common : set fact for saving topology_job_ids] ***************************
ok: [localhost]

TASK [common : Get the resource group types defined in topology] ***************
ok: [localhost]

TASK [common : Filter resource_groups by type] *********************************
ok: [localhost]

TASK [common : Assigning facts with respect to their group type] ***************
ok: [localhost]

TASK [common : Registering resource group vars from topology] ******************
skipping: [localhost]

TASK [common : output vars] ****************************************************
ok: [localhost]

PLAY [Provisioning resources based on resource group type] *********************

TASK [dummy : DEBUG:: Dummy resource group list] *******************************
skipping: [localhost]

TASK [dummy : declaring output vars] *******************************************
skipping: [localhost]

TASK [dummy : Initiating Provisioning of dummy resource group] *****************

TASK [dummy : Initiating Teardown of dummy resource group] *********************

TASK [openstack : declaring output vars] ***************************************
skipping: [localhost]

TASK [openstack : Initiating  Provisioning/Deprovioning of resources Openstack resource group] ***

TASK [aws : Provisioning aws resources] ****************************************
skipping: [localhost]

TASK [aws : declaring output vars] *********************************************
skipping: [localhost]

TASK [aws : Initiating  Provisioning of resources aws resource group] **********

TASK [aws : Initiating  teardown of resources aws resource group] **************

TASK [gcloud : DEBUG:: gcloud resource group list] *****************************
skipping: [localhost]

TASK [gcloud : declaring output vars for topology outputs of gce] **************
skipping: [localhost]

TASK [gcloud : Initiating  Provisioning of resources in gcloud resource group] *

TASK [duffy : DEBUG:: Duffy resource group list] *******************************
skipping: [localhost]

TASK [duffy : declaring output vars] *******************************************
skipping: [localhost]

TASK [duffy : Initiating Provisioning of duffy resource group] *****************

TASK [duffy : Initiating Teardown of duffy resource group] *********************

TASK [libvirt : Gather facts] **************************************************
ok: [localhost]

TASK [libvirt : declaring output vars] *****************************************
ok: [localhost]

TASK [libvirt : Initiating libvirt resource group] *****************************
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/libvirt/tasks/provision_resource_group.yml for localhost

TASK [libvirt : DEBUG:: provisioning resource group {u'resource_group_name': u'ex', u'res_group_type': u'libvirt', u'res_defs': [{u'res_name': u'linchpin-centos71', u'res_type': u'libvirt_network', u'dhcp_end': u'192.168.71.112', u'dhcp_start': u'192.168.71.101', u'ip': u'192.168.71.100'}, {u'count': 1, u'res_type': u'libvirt_node', u'uri': u'qemu:///system', u'networks': [{u'name': u'linchpin-centos71'}], u'res_name': u'centos71', u'vcpus': 1, u'memory': 1024, u'arch': u'x86_64', u'image_src': u'http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1608.qcow2.xz'}]}] ***
ok: [localhost] => {
    "msg": "The current server obj is {u'resource_group_name': u'ex', u'res_group_type': u'libvirt', u'res_defs': [{u'res_name': u'linchpin-centos71', u'res_type': u'libvirt_network', u'dhcp_end': u'192.168.71.112', u'dhcp_start': u'192.168.71.101', u'ip': u'192.168.71.100'}, {u'count': 1, u'res_type': u'libvirt_node', u'uri': u'qemu:///system', u'networks': [{u'name': u'linchpin-centos71'}], u'res_name': u'centos71', u'vcpus': 1, u'memory': 1024, u'arch': u'x86_64', u'image_src': u'http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1608.qcow2.xz'}]} "
}

TASK [libvirt : gather resource definitions of current group] ******************
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/libvirt/tasks/provision_res_defs.yml for localhost
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/libvirt/tasks/provision_res_defs.yml for localhost

TASK [libvirt : set resource_type] *********************************************
ok: [localhost]

TASK [libvirt : provision libvirt network] *************************************
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/libvirt/tasks/provision_libvirt_network.yml for localhost

TASK [libvirt : define network] ************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "The `libvirt` module is not importable. Check the requirements."}
	to retry, use: --limit @/usr/lib/python2.7/site-packages/linchpin/provision/site.retry

PLAY RECAP *********************************************************************
localhost                  : ok=16   changed=0    unreachable=0    failed=1   

hook postup initiated
[root@1f1ad66a679f lp]# echo $?
0
```

2. An openstack topology file missing **fip_pool** key fails waiting to attach a floating IP.
```
[root@cbb29dd70de8 lp]# linchpin -v --creds-path /home/lp/credentials up
ctx.workspace: /home/lp
from_api: True
calling: preup
hook preup initiated

PLAY [schema check and Pre Provisioning Activities on topology_file] ***********

TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]

TASK [common : set lp_path] ****************************************************
skipping: [localhost]

TASK [common : set lpconfig (linchpin.conf) full path] *************************
skipping: [localhost]

TASK [common : set boolean evars from linchpin.conf] ***************************
skipping: [localhost]

TASK [common : set schema evars from linchpin.conf] ****************************
skipping: [localhost]

TASK [common : set default path evars from linchpin.conf] **********************
skipping: [localhost]

TASK [common : set remaining evars from linchpin.conf] *************************
skipping: [localhost]

TASK [common : set default path evars using workspace var] *********************
skipping: [localhost]

TASK [common : if topology_file is passed, set topology to its value] **********
skipping: [localhost]

TASK [common : if schema_file is passed, set schema to its value] **************
skipping: [localhost]

TASK [common : check topology file against the provided schema] ****************
skipping: [localhost]

TASK [common : assign output from provided schema] *****************************
skipping: [localhost]

TASK [common : check topology file against schema_v3] **************************
ok: [localhost]

TASK [common : Set state for schema_v3 validation] *****************************
ok: [localhost]

TASK [common : assign schema_v3 output] ****************************************
ok: [localhost]

TASK [common : check topology file against schema_v4] **************************
skipping: [localhost]

TASK [common : update output when schema_v4 is used] ***************************
skipping: [localhost]

TASK [common : assign output from schema_v4] ***********************************
skipping: [localhost]

TASK [common : set fact for saving topology_job_ids] ***************************
ok: [localhost]

TASK [common : Get the resource group types defined in topology] ***************
ok: [localhost]

TASK [common : Filter resource_groups by type] *********************************
ok: [localhost]

TASK [common : Assigning facts with respect to their group type] ***************
ok: [localhost]

TASK [common : Registering resource group vars from topology] ******************
skipping: [localhost]

TASK [common : output vars] ****************************************************
ok: [localhost]

PLAY [Provisioning resources based on resource group type] *********************

TASK [dummy : DEBUG:: Dummy resource group list] *******************************
skipping: [localhost]

TASK [dummy : declaring output vars] *******************************************
skipping: [localhost]

TASK [dummy : Initiating Provisioning of dummy resource group] *****************

TASK [dummy : Initiating Teardown of dummy resource group] *********************

TASK [openstack : declaring output vars] ***************************************
ok: [localhost]

TASK [openstack : Initiating  Provisioning/Deprovioning of resources Openstack resource group] ***
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_resource_group.yml for localhost

TASK [openstack : Unset the authvar from previous run] *************************
ok: [localhost]

TASK [openstack : set default_cred_profile when res_grp[credentials] is undefined] ***
skipping: [localhost]

TASK [openstack : set default_cred_profile when res_grp[credentials] is undefined] ***
skipping: [localhost]

TASK [openstack : set cred_profile when res_grp[credentials] is defined] *******
ok: [localhost]

TASK [openstack : set default_cred_filename when res_grp[credentials] is defined] ***
ok: [localhost]

TASK [openstack : Get creds from auth driver] **********************************
changed: [localhost]

TASK [openstack : set auth_var] ************************************************
ok: [localhost]

TASK [openstack : provisioning resource definitions of current group] **********
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_res_defs.yml for localhost

TASK [openstack : provision resources of type os_server] ***********************
included: /usr/lib/python2.7/site-packages/linchpin/provision/roles/openstack/tasks/provision_os_server.yml for localhost

TASK [openstack : provisioning os_server resources with provided auth] *********
fatal: [localhost]: FAILED! => {"changed": false, "extra_data": null, "failed": true, "msg": "Timeout waiting for the floating IP to be attached. (Inner Exception: Error attaching IP abeef639-68d2-4342-8d8c-e4d845be3217 to server 35325ef3-5d03-47fa-97d8-c5bfdca93dd3. (404) Client Error for url: http://10.1.1.1:0000/v2.0/floatingips/abeef639-68d2-4342-8d8c-e4d845be3217.json External network 03670b37-8304-4518-9cfa-9ff17d197e12 is not reachable from subnet b403e4b4-50b7-4c2c-8f6d-f71590d71789.  Therefore, cannot associate Port aaf99f2e-d6c9-43b8-a6b3-976e38ebf8c4 with a Floating IP.)"}
	to retry, use: --limit @/usr/lib/python2.7/site-packages/linchpin/provision/site.retry

PLAY RECAP *********************************************************************
localhost                  : ok=17   changed=1    unreachable=0    failed=1   

hook postup initiated
[root@cbb29dd70de8 lp]# echo $?
0
```
My workspace files are attached. It looks to be a problem between the provider provision YAML and returning that exit code back to the main linchpin provision task YAML.
[workspace.tar.gz](https://github.com/CentOS-PaaS-SIG/linchpin/files/1171506/workspace.tar.gz)

**Comment by agharibi** [comment]
I submitted PR #344 to patch this bug. I'd appreciate some feedback if someone has the time to review it.

**Comment by herlo** [comment]
Fixed by #344
